### ISSUE TITLE
Performance improvements for transforms v2 vs. v1

### ISSUE THREAD

**Issue by pmeier** [pytorch/vision#6818]
In addition to a lot of other goodies that transforms v2 will bring, we are also actively working on improving the performance. This is a tracker / overview issue of our progress.

Performance was measured with this [benchmark script](https://gist.github.com/pmeier/e0f1ea77c9cf75b682d7f30366a89bf8). Unless noted otherwise, the performance improvements reported above were computed on uint8, RGB images and videos while running single-threaded on CPU. You can find the full benchmark results alongside the benchmark script. The results will be constantly updated if new PRs are merged that have an effect on the kernels.

Kernels: 

- color
  - [x] `adjust_brightness` #6784
  - [x] `adjust_contrast` #6784 #6933
  - [x] `adjust_gamma` #6820 #6903
  - [x] `adjust_hue` #6805 #6903 #6938
  - [x] `adjust_saturation` #6784 #6940
  - [x] `adjust_sharpness` #6784 #6930
  - [x] `autocontrast` #6811 #6935 #6942
  - [x] `equalize` #6738, #6757, #6776
  - [x] `invert` #6819
  - [x] `posterize` #6823, #6847
  - [x] `solarize` #6819
- geometry
  - [x] `affine` #6949
  - [x] `center_crop` #6880 #6949
  - [x] `crop` #6949
  - [x] `elastic` #6942
  - [x] `erase` #6983
  - [x] `five_crop`: Composite kernel #6949
  - [x] `pad` #6949
  - [x] `perspective` #6907 #6949
  - [x] `resize` #6892
  - [x] `resized_crop`: Composite kernel #6892 #6949
  - [x] `rotate` #6949
  - [x] `ten_crop`: Composite kernel #6949
- meta
  - [x] `convert_color_space` #6784 #6832
  - [x] `convert_dtype` #6795 #6903
    - There is still some performance gain left for `int` to `int` conversion. Currently, we are using a multiplication 
      but theoretically bit shifts are faster. However, on PyTorch core the CPU kernels for bit shifts are not 
      vectorized making them slower for regular sized images than a multiplication. pytorch/pytorch#88607
- misc
  - [x] `gaussian_blur` #6762 #6888
  - [x] `normalize` #6821

Transform Classes:

- [x] MixUp/CutMix #6835
- [x] ColorJitter, RandomPhotometricDistort #6837

C++ (PyTorch core):

- [x] `vertical_flip` #6983 https://github.com/pytorch/pytorch/pull/89414
- [x] `horizontal_flip` #6983 https://github.com/pytorch/pytorch/pull/88989 https://github.com/pytorch/pytorch/pull/89414

cc @vfdev-5 @datumbox @bjuncek

**Comment by datumbox** [comment]
Concerning all the methods that use affine, there are some very limited opportunities for optimization. Perhaps a couple of in-place ops in `_perspective_grid` and a few optimizations in `_apply_grid_transform` (split of mask and img, bilinear fill estimation etc). @vfdev-5 would you be OK to assess on your side whether it make sense to do these or leave the methods on `_FT` to avoid copy-pasting? Perhaps you have in mind other optimizations that I can't see that could affect performance?

**Comment by pmeier** [comment]
I did another deep dive into the ops in the second paragraph of https://github.com/pytorch/vision/issues/6818#issuecomment-1289154456 and I'm fairly confident that there is little we can do to improve on our side. The only two things I found are

- For padding modes `"edge"` and `"reflect"` we cast to `float32` and back:

  https://github.com/pytorch/vision/blob/c84dbfad97251271a789b252a2a1a52c73f623ff/torchvision/transforms/functional_tensor.py#L426-L431

  There is a long standing issue on PyTorch core pytorch/pytorch#40763 that reports this and is assigned to @vfdev-5. 

- We support `"symmetric`" padding in `F.pad`, but `torch.nn.functional.pad` doesn't. Thus, we have a custom implementation for it
  https://github.com/pytorch/vision/blob/c84dbfad97251271a789b252a2a1a52c73f623ff/torchvision/transforms/functional_tensor.py#L330
  Since it is written in Python, a possible speed up would be to implement this padding mode in C++ on the PyTorch core side.

Fixing this, we would get speed-ups for padding modes `"edge"`, `"reflect"`, and `"symmetric`" but not for the default and ubiquitous `"constant"` padding mode. Skimming the repository, it seems the only time we use non-`"constant"` padding is

https://github.com/pytorch/vision/blob/c84dbfad97251271a789b252a2a1a52c73f623ff/torchvision/transforms/functional_tensor.py#L762

In there the image is guaranteed to be float and thus would not get any performance boost. 

While I think both things mentioned above would be good to have in general, I don't think we should prioritize them.

**Comment by vfdev-5** [comment]
> Concerning elastic and all the affine transform kernels (affine, perspective, rotate), there are some very limited opportunities for optimization. Perhaps a couple of in-place ops in elastic_transform & _perspective_grid and a few optimizations in _apply_grid_transform (split of mask and img, bilinear fill estimation etc). Also some minor fixes related to the input assertion. @vfdev-5 would you be OK to assess on your side whether it make sense to do these or leave the methods on _FT to avoid copy-pasting? Perhaps you have in mind other optimizations that I can't see that could affect performance?

Checking various options with `affine`, there is no obvious way to improve runtime perfs. Yes, we can make some inplace "split of mask and img, bilinear fill estimation etc". There is also an open issue about incorrect behaviour of bilinear mode with provided not-None fill (https://github.com/pytorch/vision/issues/6517). Given that I think we can keep this implementation.

**Comment by vadimkantorov** [comment]
About not vectorized bitwise shifts, is there an issue in core?

**Comment by datumbox** [comment]
@pmeier I'm keeping the list up-to-date with all linked PRs. I'm marking as `[NEEDS RETEST]` any kernel that I touch to run further benchmarks and update the numbers.

**Comment by vadimkantorov** [comment]
An interesting question is whether a sequence of these transformations can be fused with Inductor/Dynamo and produce a fused low-memory-access CPU kernel
